vent/controller/control_module.py

```diff
@@ -443,30 +441,35 @@ def run(self):
         # start running
         # controls actuators to achieve target state
         # determined by settings and assessed by sensor values
+        self.loop_counter = 0
 
-        self.loop_counter += 1
-        now = time.time()
-        self.Balloon.update(now - self.last_update)
-        self.last_update = now
+        while self._running:
+            time.sleep(.01)
+            self.loop_counter += 1
+            now = time.time()
+            self.Balloon.update(now - self.last_update)
+            self.last_update = now
 
-        pressure = self.Balloon.get_pressure()
-        temperature = self.Balloon.temperature
+            pressure = self.Balloon.get_pressure()
+            temperature = self.Balloon.temperature
 
-        self.Controller.update(pressure)
-        Qout = self.Controller.get_Qout()
-        Qin = self.Controller.get_Qin()
-        self.Balloon.set_flow(Qin, Qout)
+            self.Controller.update(pressure)
+            Qout = self.Controller.get_Qout()
+            Qin = self.Controller.get_Qin()
+            self.Balloon.set_flow(Qin, Qout)
 
     def stop(self):
-        # stop running
-        pass
-
-    def recvUIHeartbeat(self):
-        return time.time()
+        self._running = False
```

Comment:
Does this support repetitive stop and start?

Reply:
no, it doesn't. If we want that functionality, we need multiple start and stop functions:
We would need one one start() to start up the thread, and one start() to start the controller loop. stop() would then stop the controller loop, but not the thread.

Is this the goal?

Reply:
I thought we were going to have the UICoordinator or the Main Process start the thread for the controller. And the controller just has the soft start/stop function for the loop. Why is the controller inheriting from thread?

Reply:
That would make sense. In that case yes, no need for threading in the Controller.

Zhenyu requested the change for internal threading. Not sure why. I'm just the code monkey :-)

Reply:
I've implemented the solution discussed. The class now starts a thread around the main_loop that can be controlled, repetitively, with start() and stop().